Don't presume one-to-one lookup returned an entity

gregtyler opened the pull request:
If `$this->em->find()` returns null, don't treat it like an object. Instead, just set the field to null and back out of the switch statement.

Includes a was-failing-but-now-passing test case.

Fixes #9027

gregtyler committed:
Add failing test case

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9027Test.php:
```diff
@@ -0,0 +1,77 @@
+                ]
+            );
+        } catch (Exception $e) {
+        }
```
Can you please use something similar to what was done in https://github.com/doctrine/orm/pull/8962 ?

gregtyler replied:
Done, sorry for using an old test as a template 🤦🏽

greg0ire replied:
No worries, really, I would have done the same :)

greg0ire commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH9027Test.php:
```diff
@@ -0,0 +1,77 @@
+    public function testUnitOfWorkHandlesNullRelations(): void
+    {
+        $uow = new UnitOfWork($this->_em);
+        $hints = ['fetchMode' => [GH9027Cart::class => ['customer' => ClassMetadata::FETCH_EAGER]]];
+        $cart = $uow->createEntity(GH9027Cart::class, ['id' => 1, 'customer' => 24252], $hints);
```
```suggestion
        $cart = $uow->createEntity(
                GH9027Cart::class,
                ['id' => 1, 'customer' => 24252],
                ['fetchMode' => [GH9027Cart::class => [
                        'customer' => ClassMetadata::FETCH_EAGER
                ]]];
        );
```

gregtyler replied:
I would've preferred this too, but `$hints` needs to be passed by reference, not value.

greg0ire replied:
O_o OK

greg0ire commented:
`vendor/bin/phpcbf` can help with the cs issues

gregtyler committed:
Don't presume one-to-one lookup returned an entity

If `$this->em->find()` returns null, don't treat it like an object. Instead, just set the field to null and back out of the switch statement.

Fixes #9027

gregtyler committed:
Fix formatting based on PR feedback

gregtyler commented:
@greg0ire Thanks, I've fixed the issues I could :) Let me know if there's anything else I need to do!

bendavies commented:
well this is due to
https://github.com/doctrine-extensions/DoctrineExtensions/blob/main/src/SoftDeleteable/Filter/SoftDeleteableFilter.php filtering out the select query on the FK row.

and you are working around https://github.com/doctrine-extensions/DoctrineExtensions/pull/745

if the FK exists in your row, we know an entity exists so it seems fine for the ORM to assume the data will never be null. 
i'm unsure if this fix is correct because to me, it's working around a design problem with the soft delete behavior.

gregtyler commented:
@bendavies Whilst the problem comes from SoftDeleteable in my instance, I figure we're _really_ working around a problem with Doctrine ORM Filters in general.

If I hydrate an object through `createEntity` and I have a filter which excludes an associated entity, Doctrine has to either:

1. Apply the filter and not add the associated entity
2. Ignore the filter and add the associated entity anyway
3. Throw an exception

As I understand it, Doctrine ORM's current behaviour is the first option, but it fails in PHP8 because of then mishandling the outcome if the entity is filtered out. I don't know if that's the *best* behaviour (I agree it's a silly situation to be in), but with this patch I'm just trying to keep it consistent across PHP versions.
